presto-bigquery/src/main/java/com/facebook/presto/plugin/bigquery/BigQueryPageSourceProvider.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.plugin.bigquery;
+
+import com.facebook.airlift.log.Logger;
+import com.facebook.presto.spi.ColumnHandle;
+import com.facebook.presto.spi.ConnectorPageSource;
+import com.facebook.presto.spi.ConnectorSession;
+import com.facebook.presto.spi.ConnectorSplit;
+import com.facebook.presto.spi.ConnectorTableLayoutHandle;
+import com.facebook.presto.spi.SplitContext;
+import com.facebook.presto.spi.connector.ConnectorPageSourceProvider;
+import com.facebook.presto.spi.connector.ConnectorTransactionHandle;
+import com.google.common.collect.ImmutableList;
+
+import javax.inject.Inject;
+
+import java.util.List;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+public class BigQueryPageSourceProvider
+        implements ConnectorPageSourceProvider
+{
+    private static final Logger log = Logger.get(BigQueryPageSourceProvider.class);
+    private final BigQueryStorageClientFactory bigQueryStorageClientFactory;
+    private final int maxReadRowsRetries;
+
+    @Inject
+    public BigQueryPageSourceProvider(BigQueryStorageClientFactory bigQueryStorageClientFactory, BigQueryConfig config)
+    {
+        this.bigQueryStorageClientFactory = bigQueryStorageClientFactory;
+        this.maxReadRowsRetries = config.getMaxReadRowsRetries();
+    }
+
+    @Override
+    public ConnectorPageSource createPageSource(
+            ConnectorTransactionHandle transaction,
+            ConnectorSession session,
+            ConnectorSplit split,
+            ConnectorTableLayoutHandle layout,
+            List<ColumnHandle> columns,
+            SplitContext splitContext)
+    {
+        log.debug("createPageSource(transaction=%s, session=%s, split=%s, layout=%s, columns=%s, splitContext=%s)", transaction, session, split, layout, columns, splitContext);
+        BigQuerySplit bigQuerySplit = (BigQuerySplit) split;
+        if (bigQuerySplit.representsEmptyProjection()) {
```

Comment:
could we merge the logic of BigQueryEmptyProjectionPageSource into BigQueryResultPageSource?

Reply:
BigQueryPageSourceProvider implements ConnectorPageSourceProvider, and BigQueryEmptyProjectionPageSource implements ConnectorPageSource, 2 different purpose, let's leave it for now

Reply:
the implementation is similiar to the HiveEmptySplitPageSource, I guess it's the naming is confusing, rename it to `BigQueryEmptySplitPageSource`

Reply:
we need to merge BigQueryEmptyProjectionPageSource into BigQueryResultPageSource
we do not need a separate empty page source class
BigQueryResultPageSource should cover all the logic

Reply:
same as ablove, I'd like to propose we do this logic change in another commit by a separate PR